cmd/automerge-debug/main.go

```diff
@@ -0,0 +1,942 @@
+		n >>= 7
+		if n != 0x00 {
+			b |= 0x80
+		}
+		leb = append(leb, b)
+	}
+	return leb
+}
+
+func readULEB(b []byte) (uint64, []byte, error) {
+	var n uint64
+	var i int
+	for {
+		if len(b) == 0 {
+			return 0, nil, fmt.Errorf("failed to find end of LEB")
+		}
+
+		c := b[0]
+		b = b[1:]
+		n |= uint64(c&0x7f) << (i * 7)
+		i++
+		if c&0x80 == 0 {
+			if c == 0 && i > 1 {
+				return n, b, fmt.Errorf("overly long LEB")
+			}
+			if i > 10 {
+				return n, b, fmt.Errorf("LEB > 64-bit")
+			}
+			if i == 10 && c&0x7e > 0 {
+				return n, b, fmt.Errorf("LEB > 64-bit")
+			}
+			return n, b, nil
+		}
+	}
+}
+
+func readSLEB(b []byte) (int64, []byte, error) {
+	var n uint64
+	var i int
+	var prev byte
+	for {
+		if len(b) == 0 {
+			return 0, nil, fmt.Errorf("failed to find end of LEB")
+		}
+
+		c := b[0]
+		b = b[1:]
+		n |= uint64(c&0x7f) << (i * 7)
+		i++
+		if c&0x80 == 0 {
+			if c&0x40 > 0 {
+				n |= (math.MaxUint64 << (i * 7))
+			}
+			if i > 1 && (prev&0x40 == 0 && c == 0 || prev&0x40 > 0 && c == 0x7f) {
+				return int64(n), b, fmt.Errorf("overly long LEB")
+			}
+			if i > 10 {
+				return int64(n), b, fmt.Errorf("LEB > 64-bit")
+			}
+			if i == 10 && c&0x7f != 0 && c&0x7f != 0x7f {
+				return int64(n), b, fmt.Errorf("LEB > 64-bit")
+			}
+			return int64(n), b, nil
+		}
+		prev = c
+	}
+}
+
+func addHeader(b []byte, kind byte) []byte {
+	output := append([]byte{133, 111, 74, 131, 0, 0, 0, 0, kind}, append(leb128(uint64(len(b))), b...)...)
+	hash := sha256.Sum256(output[8:])
+	output[4] = hash[0]
+	output[5] = hash[1]
+	output[6] = hash[2]
+	output[7] = hash[3]
+
+	return output
+}
+
+func main() {
+	prefix := ""
+	raw := false
+	compress := false
+	decompress := false
+	checksum := false
+
+	flag.Usage = func() {
+		fmt.Print(`usage: automerge-debug [--prefix=<doc|change>] [--raw] <file>?
+automerge-debug outputs a commented byte stream of the encoded automerge chunk.
+
+If no file is provided, the input is read from stdin.
+
+If you have the bytes of a chunk's contents, but not the header, you can pass
+--prefix=doc or --prefix=change to prefix your bytes with the correct header.
+
+The output is designed to be valid go syntax containing every input byte, you can
+edit this directly and then pass it through [bytes](github.com/ConradIrwin/bytes)
+to convert back into binary.
+`)
+
+		flag.PrintDefaults()
+	}
+	flag.StringVar(&prefix, "prefix", "", "prefix the bytes with a valid header (doc or change)")
+	flag.BoolVar(&raw, "raw", false, "output in binary")
+	flag.BoolVar(&compress, "compress", false, "compress chunk")
+	flag.BoolVar(&decompress, "decompress", false, "decompress chunk")
+	flag.BoolVar(&checksum, "fix-checksum", false, "fix checksum")
+	flag.Parse()
+
+	var file []byte
+	var err error
+	if len(flag.Args()) == 0 {
+		file, err = io.ReadAll(os.Stdin)
+	} else {
+		file, err = os.ReadFile(flag.Args()[0])
+	}
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	if strings.HasPrefix(prefix, "doc") {
```

Comment:
`switch-case` can be here 😉

Reply:
Thanks! I've left this as if-statements because (to my personal taste at least) it's clearer the way it is right now. I think that would change if each branch was testing a value, or was at least more similar.